witness/tests/Test/Dscp/Witness/Tx/MoneyTx.hs

```diff
@@ -106,26 +106,28 @@ spec_Money_tx_application = do
             nonce <- pick $ arbitrary `suchThat` (/= 0)
             let steps = properSteps & tcsInAcc .~ \addr -> TxInAcc addr nonce
             let tx = makeTx steps txData
-            lift $ throwsPrism (_AccountError . _NonceMustBeIncremented) $
+            lift $ throwsPrism (_AccountError . _NonceMismatch) $
                 applyTx tx
 
-        it "Bad input address is not fine" $ witnessProperty $ do
+        it "Input address different from signer is not fine" $ witnessProperty $ do
             txData <- pick genSafeTxData
-            sourceAddr <- pick (arbitrary @Address)
+            sourceAddr <- mkAddr . toPublic <$> pick selectGenesisSecret
             pre (sourceAddr /= mkAddr (toPublic $ tdSecret txData))
 
             let steps = properSteps
                     & tcsInAcc %~ \f addr -> (f addr){ tiaAddr = sourceAddr }
             let tx = makeTx steps txData
-            lift $ throwsSome $ applyTx tx
+
+            lift $ throwsPrism (_AccountError . _WitnessMismatchesInput) $
+                applyTx tx
 
         it "Having no money is not fine" $ witnessProperty $ do
             txData' <- pick genSafeTxData
             secret <- pick $ arbitrary `suchThat` (`notElem` testGenesisSecrets)
             let txData = txData'{ tdSecret = secret }
 
             let tx = makeTx properSteps txData
-            lift $ throwsPrism (_AccountError . _BalanceCannotBecomeNegative) $
+            lift $ throwsPrism (_AccountError . _InsufficientBalance) $
```

Comment:
What does `throwsPrism` do? Repacks exception with a prism? Or does it checks that only exception, covered by prism, is thrown? Why not `assertThrowsPrism`?

Reply:
Also, is it possible to remove all these `lift`s? We have nice test logic here, its a pity that all this plumbing is sticking out of the wall.

Reply:
> Or does it checks that only exception, covered by prism, is thrown? 

Yep. Since the verb ends with `s`, I thought this function looks like a predicate.
I would rename, but this function has been used in too many places at now.


> Also, is it possible to remove all these lifts? We have nice test logic here, its a pity that all this plumbing is sticking out of the wall.

I tried really hard to avoid these `lift`s, and it looks like the best we can live with.

The approach you used in `sqliteProperty`, where arbitrarily-generated parameter was passed to a test case, scales not very well: I cannot use custom generators (like `genSafeTxData` above) without some half-working hacks, and writing preconditions (like with `pre` function) makes us use `if/then/else` with the corresponding indentation madness.

Thus I use `PropertyM` mostly everywhere for testing monadic code, but implementing `MonadUnliftIO` for such Cont-based monad seems impossible, so `lift`s.